### PR TITLE
Mini Project JQ

### PR DESCRIPTION
**Pull request by hazel-sudz**
note: submitted with automatic late extension

# Structure Overview
- Map (map(), command_match())
    - add
    - array_index
    - array_slice
    - del_array
    - del_key
    - identity
    - length
    - object_identifier
 - args
 - formatter
 - parser
 - processor
 - main

# Example Colored Output
Screenshot 2024-10-19 at 6 08 55 PM

# Example Monochrome Output
Screenshot 2024-10-19 at 6 58 18 PM

# Example Compact Output
Screenshot 2024-10-19 at 6 10 17 PM

# Example Custom Indent
Screenshot 2024-10-19 at 6 11 32 PM

# Example Handling Conflicting Arguments
Screenshot 2024-10-19 at 6 10 58 PM

# Example Array Index
Screenshot 2024-10-19 at 6 13 10 PM

# Example Array Slice
Screenshot 2024-10-19 at 6 13 33 PM

# Example Pipe
Screenshot 2024-10-19 at 6 14 14 PM

# Example Array Iterator
Screenshot 2024-10-19 at 6 41 08 PM

# Example Add
Screenshot 2024-10-19 at 6 42 05 PM

# Example Length
Screenshot 2024-10-19 at 6 42 28 PM

# Example Del
Screenshot 2024-10-19 at 6 43 10 PM

 
# Comprehensive Testing
Screenshot 2024-10-19 at 6 45 17 PM

### CODE REVIEW

File: mini-projects/00-jq-rs/src/formatter.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    Ok(format!("\x1b[{color_code}m{input}\x1b[0m"))
+}
+
+// given a JSON value represented by serde_json::Value, formats it according to the JQ manual
+// given the recursive structure of JSON we approach this problem recursively
+// support provided for --sort-keys, --indent, --compact-output, and --monochrome-output flags
+// indent levels are pushed/poped as we recurse into the JSON structure
+pub fn format(
+    input: Value,
+    sort_keys: bool,
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/formatter.rs:53]: I'd consider defining an options struct for the argument of this function to make life easier on the calling end

File: mini-projects/00-jq-rs/src/args.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    #[clap(long, action, conflicts_with = "color_output")]
+    pub monochrome_output: bool,
+
+    #[clap(long, action)]
+    pub sort_keys: bool,
+
+    #[clap(long, conflicts_with = "compact_output")]
+    pub indent: Option<u8>,
+
+    #[clap(long, action, conflicts_with = "indent")]
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/args.rs:22]: Ooh, I didn't even know about `conflicts_with` - nice find!

File: mini-projects/00-jq-rs/src/maps/del_array.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
+use crate::maps::map::Map;
+use anyhow::Result;
+use regex::Regex;
+use serde_json::Value;
+
+// implements the "del(.[from:to])" operation
+// deletes the slice of the array from the given from index to the given to index
+// returns the remaining elements of the array
+// returns an error if the value is not an array
+pub struct DelMapArray {
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/maps/del_array.rs:10]: Great pattern here of using traits and storing implementation specific parameters as struct members throughout

File: mini-projects/00-jq-rs/src/maps/del_array.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    // replicates echo '[0,1,2]' | jq "del(.[0,1])"
+    #[test]
+    fn test_basic_del_array() {
+        let del_map = DelMapArray { from: 0, to: 1 };
+        let values = del_map
+            .map(Ok(vec![serde_json::from_str("[0,1,2]").unwrap()]))
+            .unwrap();
+        assert_eq!(values.len(), 1);
+        assert_eq!(values[0].to_string(), "[1,2]");
+    }
```
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/src/maps/del_array.rs:65]: could do some some out of range tests here (you might have found that `drain` panics, which may or may not be your intended behavior)

File: mini-projects/00-jq-rs/tests/integration.rs
Comment (amit-kumarh) [inline mini-projects/00-jq-rs/tests/integration.rs:1]: 😲 Integration tests?? neato

**Review by amit-kumarh — COMMENTED** [review]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points

**Submission Score: 4/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

Great job! Some operations could have a few more test cases for thoroughness' sake (eg out-of-range deletion attempts for DelMapArray), but just minor nits.

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

### Points
**Features Score: 61/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

Nice job! In particular, the method you used to not have to change your `parser.rs` if you add another filter is quite neat (if a little inefficient)

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

Would be nice to see some more documentation and particularly use of RustDoc / documentation of how to expand your implementation, but the module level comments are appreciated. (-0.5)

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 15/15**

## Summary
Add up the point totals here

**Total Score: 79.5/80**
